doc/api/module.md

```diff
@@ -269,65 +164,224 @@ changes:
+
+// Because this is a dynamic `import()`, the `http-to-https` hooks will run
+// to handle `./my-app.js` and any other files it imports or requires.
+await import('./my-app.js');
+```
+
+```cjs
+const { register } = require('node:module');
+const { pathToFileURL } = require('node:url');
+
+register('http-to-https', pathToFileURL(__filename));
+
+// Because this is a dynamic `import()`, the `http-to-https` hooks will run
+// to handle `./my-app.js` and any other files it imports or requires.
+import('./my-app.js');
+```
+
+In this example, we are registering the `http-to-https` hooks, but they will
+only be available for subsequently imported modules—in this case, `my-app.js`
+and anything it references via `import` or `require`. If the
+`import('./my-app.js')` had instead been a static `import './my-app.js'`, _the
+app would already have been loaded_ before the `http-to-https` hooks were
+registered. This is part of the design of ES modules, where static imports are
+evaluated from the leaves of the tree first back to the trunk. There can be
+static imports _within_ `my-app.js`, which will not be evaluated until
+`my-app.js` is when it's dynamically imported.
 
-When hooks are used they apply to each subsequent loader, the entry point, and
-all `import` calls. They won't apply to `require` calls; those still follow
-[CommonJS][] rules.
+The `my-app.js` file can also be CommonJS. Customization hooks will run for any
+modules that it references via either `import` or `require`.
 
-Loaders follow the pattern of `--require`:
+Finally, if all you want to do is register hooks before your app runs and you
+don't want to create a separate file for that purpose, you can pass a `data:`
+URL to `--import`:
 
 ```bash
-node \
-  --experimental-loader unpkg \
-  --experimental-loader http-to-https \
-  --experimental-loader cache-buster
+node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("http-to-https", pathToFileURL("./").href);' ./my-app.js
 ```
 
-These are called in the following sequence: `cache-buster` calls
-`http-to-https` which calls `unpkg`.
+### Chaining
+
+It's possible to call `register` more than once:
+
+```mjs
+// entrypoint.mjs
+import { register } from 'node:module';
+
+register(new URL('./first.mjs', import.meta.url));
+register('./second.mjs', import.meta.url);
+await import('./my-app.mjs');
+```
+
+```cjs
+// entrypoint.cjs
+const { register } = require('node:module');
+const { pathToFileURL } = require('node:url');
+
+const parentURL = pathToFileURL(__filename);
+register(new URL('./first.mjs', parentURL));
+register('./second.mjs', parentURL);
+import('./my-app.mjs');
+```
+
+In this example, the registered hooks will form chains. If both `first.mjs` and
+`second.mjs` define a `resolve` hook, both will be called, in the order they
+were registered. The same applies to all the other hooks.
+
+### Communication between main and hooks threads
```

Comment:
Should we keep calling it the loader thread?

Reply:
I don't think so. It's the thread where the hooks run. One term is better than two.

Reply:
Orchestrator/orchestration threads?

Reply:
Customization thread?

I wasn't assuming we'd need a thread like this (or use this one) for other systems, but who knows?